### PR TITLE
Require hashes for dependencies

### PR DESCRIPTION
**Pull request by sgillies**
I'm turning mercantile into an example of checking hashes of runtime and testing dependencies based on what James Bennett wrote in https://www.b-list.org/weblog/2018/apr/25/lets-talk-about-packages/.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,125 @@
...
+typing-extensions==3.7.4.3 \
+    --hash=sha256:7cb407020f00f7bfc3cb3e7881628838e69d8f3fcab2f64742a5e76b2f841918 \
+    --hash=sha256:99d4073b617d30288f569d3f13d2bd7548c3a7e4c8de87db09a9d29bb3a4a60c \
+    --hash=sha256:dafc7639cde7f1b6e1acc0f457842a83e722ccca8eef5270af2d74792619a89f
+urllib3==1.26.4 \
+    --hash=sha256:2f4da4594db7e1e110a944bb1b551fdf4e6c136ad42e4234131391e21eb5b0df \
+    --hash=sha256:e7b021f7241115872f92f43c6508082facffbd1c048e3c6e2bb9c2a157e28937
+zipp==3.4.1 \
+    --hash=sha256:3607921face881ba3e026887d8150cca609d517579abe052ac81fc5aeffdbd76 \
+    --hash=sha256:51cb66cc54621609dd593d1787f286ee42a5c0adbb4b29abea5a63edc3e03098
```
Comment (sgillies) [inline requirements.txt:125]: This new file was generated in this way:

1. Creating a new Python environment
2. `pip install ...`
3. `pip freeze > requirements.txt`
4. `cat requirements.txt | while read line; do hashin -a sha256 $line; done`

Following directions in https://www.b-list.org/weblog/2018/apr/25/lets-talk-about-packages/. Certainly many of these hashes will never apply, but do no harm here.

File: setup.py [hunk 2]
```diff
@@ -38,7 +38,7 @@
     install_requires=["click>=3.0"],
     extras_require={
         "dev": ["check-manifest"],
-        "test": ["coveralls", "hypothesis", "pytest-cov", "pydocstyle"],
+        "test": ["hypothesis", "pytest"],
```
Comment (sgillies) [inline setup.py:41]: Tests for users don't depend on coveralls or pydocstyle.

File: requirements.txt (hunk 1 above)
Comment (daniel-j-h) [reply requirements.txt:125]: Quick question, have you ever looked into pip-tools

https://github.com/jazzband/pip-tools

that's what I have used in the past for getting hashes and reproducible deps. going with pip.

Sidenote: I've switched to poetry by now completely and I'm very happy for 99% of the use cases I have so far; wondering if you have opinions on that, or what benefits you see in doing all these manual hacks on top of pip where poetry does it The Right Way :tm: from the start?

Comment (sgillies) [reply requirements.txt:125]: I have used pip-tools before, but for some reason it didn't stick. I might have been trying too many other new things at the same time.

I don't see what poetry offers this project. The requirements.txt file is only intended for use with our CI and I don't plan to update it very often.

Comment (daniel-j-h) [reply requirements.txt:125]: Gotcha :raised_hands: 

I've used poetry for libraries as well, by just providing the `pyproject.toml` and letting `poetry build` come up with files like `setup.py` for me.

**Pull request merged**